creusot-contracts/src/ghost/ptr_own.rs

```diff
@@ -234,17 +228,174 @@ impl<T: ?Sized> PtrOwn<T> {
+/// - To extract a [`&PtrLive<T>`][PtrLive] (evidence used by pointer arithmetic): [`live`](PtrOwn::live), [`live_mut`](PtrOwn::live_mut).
+impl<T> PtrOwn<[T]> {
+    /// The number of elements in the slice.
+    #[logic(open, inline)]
+    pub fn len(&self) -> Int {
+        pearlite! { self.val()@.len() }
+    }
+
+    /// Split a `&PtrOwn<[T]>` into two subslices of lengths `index` and `self.len() - index`.
+    #[trusted]
+    #[check(ghost)]
+    #[requires(0 <= index && index <= self.len())]
+    #[ensures(self.ptr().thin() == result.0.ptr().thin() && self.ptr().thin().offset_logic(index) == result.1.ptr().thin())]
+    #[ensures(index == result.0.len() && self.len() - index == result.1.len())]
+    #[ensures(forall<k: Int> 0 <= k && k < index ==> self.val()@[k] == result.0.val()@[k])]
+    #[ensures(forall<k: Int> index <= k && k < self.len() ==> self.val()@[k] == result.1.val()@[k - index])]
+    pub fn split_at_ghost(&self, index: Int) -> (&Self, &Self) {
+        let _ = index;
+        panic!("called ghost function in normal code")
+    }
+
+    /// Split a `&mut PtrOwn<[T]>` into two subslices of lengths `index` and `self.len() - index`.
+    #[trusted]
+    #[check(ghost)]
+    #[requires(0 <= index && index <= self.len())]
+    #[ensures(self.ptr().thin() == result.0.ptr().thin() && self.ptr().thin().offset_logic(index) == result.1.ptr().thin())]
+    #[ensures(index == result.0.len() && self.len() - index == result.1.len())]
+    #[ensures(index == (^result.0).len() && self.len() - index == (^result.1).len())]
+    #[ensures(forall<k: Int> 0 <= k && k < index ==> self.val()@[k] == result.0.val()@[k] && (^self).val()@[k] == (^result.0).val()@[k])]
+    #[ensures(forall<k: Int> index <= k && k < self.len() ==> self.val()@[k] == result.1.val()@[k - index] && (^self).val()@[k] == (^result.1).val()@[k - index])]
+    #[ensures((^self).ptr() == self.ptr())]
+    pub fn split_at_mut_ghost(&mut self, index: Int) -> (&mut Self, &mut Self) {
+        let _ = index;
+        panic!("called ghost function in normal code")
+    }
+
+    /// Split `&PtrOwn<[T]>` into a sequence of `&PtrOwn<T>` for each element.
+    #[trusted]
+    #[check(ghost)]
+    #[ensures(result.len() == self.len())]
+    #[ensures(forall<i> 0 <= i && i < self.len()
+        ==> result[i].ptr() == self.ptr().thin().offset_logic(i)
+        && *result[i].val() == self.val()@[i])]
+    pub fn elements(&self) -> Seq<&PtrOwn<T>> {
+        panic!("called ghost function in normal code")
+    }
+
+    /// Split `&mut PtrOwn<[T]>` into a sequence of `&mut PtrOwn<T>` for each element.
+    #[trusted]
+    #[check(ghost)]
+    #[ensures(result.len() == self.len())]
+    #[ensures(forall<i> 0 <= i && i < self.len()
+        ==> result[i].ptr() == self.ptr().thin().offset_logic(i)
+        && *result[i].val() == self.val()@[i]
+        && (^self).ptr() == self.ptr()
+        && *(^result[i]).val() == (^self).val()@[i])]
```

Comment:
```suggestion
    #[ensures(forall<i> 0 <= i && i < self.len()
        ==> result[i].ptr() == self.ptr().thin().offset_logic(i)
        && *result[i].val() == self.val()@[i]]
    #[ensures(forall<i> 0 <= i && i < self.len()
        && (^self).ptr() == self.ptr()
        ==> *(^result[i]).val() == (^self).val()@[i])]
```

Otherwise, it's unsound (one can write into one of the returned borrow.

Reply:
```suggestion
    #[ensures(forall<i> 0 <= i && i < self.len()
        ==> result[i].ptr() == self.ptr().thin().offset_logic(i)
        && *result[i].val() == self.val()@[i])]
    #[ensures((forall<i> 0 <= i && i < self.len() && result[i].ptr() == (^result[i]).ptr())
        ==> (^self).ptr() == self.ptr()
        && forall<i> 0 <= i && i < self.len() *(^result[i]).val() == (^self).val()@[i])]
```

(Surprisingly subtle...)

Reply:
Yes. I have to say that this is an idea that I just had I have not analyzed all the implications. If that works, then it should be used for resource algebra's split_mut, where we essentially have the same problem. Cc @arnaudgolfouse.